Add flexible VNet and subnet deployment support

keshava-kandikonda opened the pull request:
This PR adds flexible VNet and subnet deployment support with three scenarios:

- New VNet + new subnets - Creates everything from scratch
- Existing VNet + new subnets - Creates subnets in existing VNet
- Existing VNet + existing subnets - Adds delegations to existing subnets without changing address ranges
- 

Changes

- Added parameters: useExistingVnet, createSubnets, vnetAddressPrefix
- Added variables: subnetAddressForApp, subnetAddressForDb
- Added add_delegations deployment that uses reference() to preserve existing subnet address ranges
- Removed CDN references (deprecated)

Commit:
Add flexible VNet and subnet deployment support

- Support three scenarios: new VNet, existing VNet with new subnets, existing VNet with existing subnets
- Add parameters: useExistingVnet, createSubnets, vnetAddressPrefix, subnet addresses
- Add update_existing_subnets nested deployment for subnet delegation updates
- Remove hardcoded resource group/subscription IDs
- Remove CDN-related resources

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,48 @@
+                "description": "Set to false to skip subnet creation (useful when subnets already exist)"
+            }
+        },
+        "vnetAddressPrefix": {
```
Do we need this ?

keshava-kandikonda replied:
Fixed this- Updated parameter description to clarify this only applies when creating a new VNET.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,48 @@
+                "description": "Address space for the VNet (e.g., 10.0.0.0/16)"
+            }
+        },
+        "subnetAddressForApp": {
```
Same question..

For new, we create the ranges. If the subnet exists we just reuse and don't update the ranges right?

keshava-kandikonda replied:
Fixed - Updated parameter descriptions to clarify these only apply when createSubnets=true. For existing subnets, their address ranges are preserved via reference() function. Validated with test deployment.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,48 @@
+                "description": "Address prefix for the Database subnet (e.g., 10.0.1.0/24)"
+            }
+        },
+        "useExistingPrivateDnsZone": {
```
Are we using it?

keshava-kandikonda replied:
Removed it

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,48 @@
         "managedIdentityName": {
             "type": "string"
+        },
+        "useExistingVnet": {
```
Instead of this, Let's make a the param like VnetResourceId and take the decision to create a new Vnet if it is empty?

vkarasala replied:
Let's make a variable internally based on VnetResourceId  and reuse at places.

keshava-kandikonda replied:
The vnetResourceId approach would anyway require keeping vnetName (for creating new VNets), creating parameter overlap. The boolean flag would be clearer: useExistingVnet=false uses vnetName to create new, useExistingVnet=true uses vnetName+vnetResourceGroup to reference existing.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,48 @@
+                "description": "Set to true to use an existing VNet instead of creating a new one"
+            }
+        },
+        "createSubnets": {
```
Instead of this, it should be appSubnetName, dbSubnetName. If they are not empty, let's add just the delegation and otherwise delegation + AddressRange i.e createNewSubnet

keshava-kandikonda replied:
It may get more confusing for cx to leave subnet parameters empty to create new subnets right?

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -543,11 +566,11 @@
             ],
-            "resourceGroup": "rajneesh-experiments",
-            "subscriptionId": "9801ce74-29e4-4846-9d02-5010413d1042",
+            "resourceGroup": "[parameters('serverFarmResourceGroup')]",
```
Do we need this? I believe it picks up the defaults from the context we run??

keshava-kandikonda replied:
Do we go with the assumption that cx will use the same resource group? Also, should we keep hardcoded RG and subscription like the original code in the actual template?

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -532,7 +555,7 @@
                 "addressSpace": {
                     "addressPrefixes": [
-                        "[variables('vnetAddress')]"
+                        "[parameters('vnetAddressPrefix')]"
```
Let's revert this, If we are creating new vnet, let's not extend the options and keep it untouched for now.

Commit:
Preserve existing subnet ranges and cleanup

Use reference() function to read existing subnet addressPrefix instead of omitting it.
Remove unused useExistingPrivateDnsZone parameter and update documentation.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -208,6 +196,34 @@
+        },
+        "subnetAddressForApp": {
+            "type": "string",
+            "defaultValue": "10.0.0.0/24",
```
Why do we want to expose them as param is the question, For portal we don't do that. Can we keep it as variable and use it when necessary. I really want to reduce the number of parameters which users don't change for most of the cases.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -595,6 +602,60 @@
+                            "apiVersion": "2021-03-01",
+                            "name": "[format('{0}/{1}', parameters('vnetName'), parameters('subnetForApp'))]",
+                            "properties": {
+                                "addressPrefix": "[reference(resourceId(parameters('serverFarmResourceGroup'), 'Microsoft.Network/virtualNetworks/subnets', parameters('vnetName'), parameters('subnetForApp')), '2021-03-01').addressPrefix]",
```
without addressPrefx we run into issues?

keshava-kandikonda replied:
Yes, we are running into issues without it.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -595,6 +602,60 @@
         },
+        {
+            "type": "Microsoft.Resources/deployments",
+            "name": "update_existing_subnets",
```
nit: add_delegations

vkarasala commented on ARM_Template_Create_WP/azuredeploy.parameters.json:
```diff
@@ -208,6 +196,18 @@
+            "value": true
+        },
+        "subnetAddressForApp": {
+            "value": "10.0.0.0/24"
```
need these?

vkarasala commented:
LGTM, but please add some testing evidence demonstrating the default flow and existing vnet/subnet flows.

vkarasala commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -595,6 +588,60 @@
+                            "apiVersion": "2021-03-01",
+                            "name": "[format('{0}/{1}', parameters('vnetName'), parameters('subnetForApp'))]",
+                            "properties": {
+                                "addressPrefix": "[reference(resourceId(parameters('serverFarmResourceGroup'), 'Microsoft.Network/virtualNetworks/subnets', parameters('vnetName'), parameters('subnetForApp')), '2021-03-01').addressPrefix]",
```
I see the only difference between this and subnet_resoruces is the addressPrefix, Can we add the condition and remove the duplication?

siriande replied:
What if the VNet is in a different resource group?

siriande commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -405,7 +399,7 @@
                 "[resourceId('Microsoft.Network/privateDnsZones/virtualNetworkLinks', parameters('privateDnsZoneNameForDb'), format('{0}-vnetlink', parameters('privateDnsZoneNameForDb')))]",
                 "[concat('Microsoft.ManagedIdentity/userAssignedIdentities/', parameters('managedIdentityName'))]",
-                "[resourceId('Microsoft.Resources/deployments', 'subnet_Resources')]"
+                "[if(parameters('createSubnets'), resourceId('Microsoft.Resources/deployments', 'subnet_Resources'), if(parameters('useExistingVnet'), resourceId('Microsoft.Resources/deployments', 'add_delegations'), resourceId('Microsoft.Network/virtualNetworks/', parameters('vnetName'))))]"
```
Please correct me if I'm wrong. Souldn't the default dependency be `subnet_Resources`? If both the custom flags are false, it means to create a new VNet with subnets. So, this should wait until the subnets are ready.

siriande commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -622,7 +669,7 @@
             "dependsOn": [
                 "[resourceId('Microsoft.Network/virtualNetworks/', parameters('vnetName'))]",
-                "[resourceId('Microsoft.Resources/deployments', 'subnet_Resources')]"
+                "[if(parameters('createSubnets'), resourceId('Microsoft.Resources/deployments', 'subnet_Resources'), if(parameters('useExistingVnet'), resourceId('Microsoft.Resources/deployments', 'add_delegations'), resourceId('Microsoft.Network/virtualNetworks/', parameters('vnetName'))))]"
```
same here. Maybe reordering this would make this simpler:
```
if (not useExistingVnet) or (createSubnets) -> resourceId('Microsoft.Resources/deployments', 'subnet_Resources')
else -> resourceId('Microsoft.Resources/deployments', 'add_delegations')
```

siriande commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -543,11 +538,9 @@
             "type": "Microsoft.Resources/deployments",
             "name": "subnet_Resources",
             "apiVersion": "[variables('resourceDeploymentApiVersion')]",
+            "condition": "[parameters('createSubnets')]",
```
Should we do `useExistingVnet is false or createSubnets is true` just in case the defaults are overridden?

Commit:
Add vnetresourceGroup parameter

siriande commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -543,11 +545,9 @@
             "apiVersion": "[variables('resourceDeploymentApiVersion')]",
+            "condition": "[or(not(parameters('useExistingVnet')), parameters('createSubnets'))]",
             "dependsOn": [
-                "[resourceId('Microsoft.Network/virtualNetworks/', parameters('vnetName'))]"
```
This should wait for the VNet to be created when provisioning new VNet resource i.e., `useExistingVnet` is `false`.

siriande commented on ARM_Template_Create_WP/azuredeploy.json:
```diff
@@ -543,11 +545,9 @@
             "type": "Microsoft.Resources/deployments",
             "name": "subnet_Resources",
```
The template might try to create the subnets in the Resource Group where the deployment is running. Can you verify this scenario: _Use existing VNet from a different RG and create new subnet resources_ ?

Commit:
Fix cross RG scenarios

siriande reviewed (commented):
Could you also add all the validated test cases in the PR description?
